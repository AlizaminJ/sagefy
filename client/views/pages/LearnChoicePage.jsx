/* eslint-disable camelcase */
import React from 'react'
import ReactMarkdown from 'react-markdown'
import { string, number, shape, objectOf, bool } from 'prop-types'
import { shuffle } from 'shuffle-seed'
import Layout from '../components/Layout'
import Icon from '../components/Icon'

export default function LearnChoicePage({
  hash,
  progress,
  card: {
    name,
    data: { body, max_options_to_show, options },
  },
  body: {
    choice,
    seed = process.env.NODE_ENV === 'test' ? '_' : Date.now().toString(36),
  } = {},
}) {
  const xoptions = shuffle(Object.entries(options), seed).slice(
    0,
    max_options_to_show
  )
  return (
    <Layout hash={hash} page="LearnChoicePage" title="Learn" description="-">
      {progress && (
        <section>
          <progress value={progress} />
        </section>
      )}

      <form action={choice ? '/next' : ''} method={choice ? 'GET' : 'POST'}>
        <section>
          <div alt={name}>
            <ReactMarkdown source={body} />
          </div>
          <ul className="ls-n">
            {xoptions.map(([key, { value, correct, feedback }]) => (
              <li key={key}>
                <input
                  type="radio"
                  name="choice"
                  value={key}
                  id={key}
                  disabled={choice}
                  checked={choice === key}
                />{' '}
                <label htmlFor={key} className="d-i">
                  {value}
                  {choice === key && <br />}
                  {choice === key && (
                    <mark className={correct ? 'good' : ''}>
                      {correct ? <Icon i="check" /> : <Icon i="error" />}{' '}
                      {feedback}
                    </mark>
                  )}
                </label>
              </li>
            ))}
            <input type="hidden" name="seed" value={seed} />
          </ul>
        </section>

        <section>
          {choice ? (
            <button type="submit">
              <Icon i="card" /> Next Card
            </button>
          ) : (
            <button type="submit">
              <Icon i="check" /> Check Answer
            </button>
          )}
        </section>
      </form>
    </Layout>
  )
}

LearnChoicePage.propTypes = {
  hash: string.isRequired,
  progress: number,
  card: shape({
    name: string.isRequired,
    data: shape({
      body: string.isRequired,
      max_options_to_show: number.isRequired,
      options: objectOf(
        shape({
          value: string.isRequired,
          correct: bool.isRequired,
          feedback: string.isRequired,
        })
      ),
    }).isRequired,
  }).isRequired,
  body: shape({
    choice: string,
    seed: string,
  }),
}

LearnChoicePage.defaultProps = {
  progress: null,
  body: {
    choice: null,
    seed: undefined,
  },
}
