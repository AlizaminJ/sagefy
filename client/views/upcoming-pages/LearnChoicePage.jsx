import React from 'react'
import Icon from '../components/Icon'
// TODO import ReactMarkdown from 'react-markdown'

export default function LearnChoicePage() {
  return (
    <div className="LearnChoicePage">
      <section>
        <progress value={0.7} />
      </section>

      <section>
        <div>
          <p>What is the difference between harmonic and inharmonic waves?</p>
        </div>
        <ul className="ls-n">
          <li>
            <input
              type="radio"
              name="choice"
              value="F0-61DPpSxqURHvryZy84Q"
              id="F0-61DPpSxqURHvryZy84Q"
              disabled
            />{' '}
            <label htmlFor="F0-61DPpSxqURHvryZy84Q" className="d-i">
              Both harmonics and inharmonics are integer multiples of the
              fundamental.
            </label>
          </li>
          <li>
            <input
              type="radio"
              name="choice"
              value="TT8B9zrBRkq5maek13ASJg"
              id="TT8B9zrBRkq5maek13ASJg"
              disabled
            />{' '}
            <label htmlFor="TT8B9zrBRkq5maek13ASJg" className="d-i">
              Inharmonics are integer multiples, harmonics are not.
            </label>
          </li>
          <li>
            <input
              type="radio"
              name="choice"
              value="R8eK8S_xQG2D6iJku1JS9Q"
              id="R8eK8S_xQG2D6iJku1JS9Q"
              disabled
              checked
            />{' '}
            <mark>
              <Icon i="check" /> Good job!
            </mark>{' '}
            <label htmlFor="R8eK8S_xQG2D6iJku1JS9Q" className="d-i">
              Harmonics are integer multiples, inharmonics are not.
            </label>
          </li>
          <li>
            <input
              type="radio"
              name="choice"
              value="lXQx9AQSQjKfSfIYy2OqWw"
              id="lXQx9AQSQjKfSfIYy2OqWw"
              disabled
            />{' '}
            <label htmlFor="lXQx9AQSQjKfSfIYy2OqWw" className="d-i">
              Neither harmonics nor inharmonics are integer multiples of the
              fundamental.
            </label>
          </li>
        </ul>
      </section>

      <section>
        <form action="/learn-page/1">
          <button type="submit">
            <Icon i="card" /> Next Card
          </button>
          {/* or, Check Answer */}
        </form>
      </section>
    </div>
  )
}
