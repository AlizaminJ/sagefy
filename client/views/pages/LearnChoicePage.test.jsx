import LearnChoicePage from './LearnChoicePage'

describe('LearnChoicePage', () => {
  it('should render LearnChoicePage', () => {
    expect(LearnChoicePage({ data: { options: {} } })).toMatchSnapshot()
  })
})
