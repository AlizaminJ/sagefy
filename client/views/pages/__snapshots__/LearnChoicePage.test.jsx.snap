// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LearnChoicePage should render LearnChoicePage 1`] = `
<Layout hash={hash}
  description="-"
  page="LearnChoicePage"
  title="Learn"
>
  <form
    action=""
    method="POST"
  >
    <section>
      <div>
        <ReactMarkdown
          astPlugins={Array []}
          escapeHtml={true}
          plugins={Array []}
          rawSourcePos={false}
          renderers={Object {}}
          skipHtml={false}
          sourcePos={false}
          transformLinkUri={[Function]}
        />
      </div>
      <ul
        className="ls-n"
      >
        <input
          name="seed"
          type="hidden"
          value="_"
        />
      </ul>
    </section>
    <section>
      <button
        type="submit"
      >
        <Icon
          i="check"
          s="m"
        />
         Check Answer
      </button>
    </section>
  </form>
</Layout>
`;
